PlayerSkeleton.java

```diff
@@ -1,22 +1,217 @@
 
+import java.util.Arrays;
+
 public class PlayerSkeleton {
 
+	private int[][] pWidth;
+	private int[][] pHeight;
+	private int[][][] pBottom;
+	private int[][][] pTop;
+
 	//implement this function to have a working system
 	public int pickMove(State s, int[][] legalMoves) {
-		// System.out.println("Heuristic score = " + Heuristic.evaluate(s));
-		return 0;
+		
+		//Step 1. Get a copy of the current board configuration.
+		int[][] currentField = copyArray(s.getField());
+
+		//Step 2. Test all possible moves of the piece on the board and choose the move with the best heuristic score 
+		int best = tryPossibleMoves(currentField, legalMoves, s);
+
+		return best;
+	}
+
+	/**
+	* This method test all possible moves of the piece and returns the move with the best heuristic score
+	* @param currentField (current board config), legalMoves (possible moves for current piece), s (current board configuration)
+	* @return move with best (highest) heuristic score
+	**/
+	private int tryPossibleMoves(int[][] currentField, int[][] legalMoves, State s) {
+		//legalMoves = all possible moves of the current piece.
+		//Step 1. Apply each move (action) to get the change field (board configuration).
+		//Step 2. Apply Heuristic.evaluate(s) to get the heuristic score from the move.
+		//Step 3. Get the best move.
+		int bestScore = 0;
+		int bestMove = 0;
+
+		for(int i=0; i<legalMoves.length; i++) {
+			//Step 1
+			int [][] field = copyArray(currentField);
+			int[] top = copyArray(s.getTop());
+			Pair results = testMove(legalMoves[i], field, s.getNextPiece(), top);
+			//Step 2
+			if(results == null) {
+				continue;
+			}
+			int score = Heuristic.evaluate(results.getField(), results.getRowsCleared());
+			//System.out.println("Heuristic score = " + Heuristic.evaluate(s));
+
+			//Step 3
+			if(score > bestScore) {
+				bestScore = score;
+				bestMove = i;
+			}
+
+		}
+		return bestMove;
+	}
+
+	/**
+	* This method gets the change field from testing the possible move of a piece. 
+	* @param move (action), field (current board configuration), currentPiece (current piece to place), top (?)
+	* @return new board configuration after applying the move (action)
+	**/
+	private Pair testMove(int[] move, int[][] field, int currentPiece, int[] top) {
+		int orient = move[0];
+		int slot = move[1];
+		int row = 21;
+		int col = 10;
+
+		//height if the first column makes contact
+		int height = top[slot]-pBottom[currentPiece][orient][0];
+		//for each column beyond the first in the piece
+		for(int c = 1; c < pWidth[currentPiece][orient];c++) {
+			height = Math.max(height,top[slot+c]-pBottom[currentPiece][orient][c]);
+		}
+		
+		//Check if move is valid
```

Comment:
Oh but I thought the legal moves are already given to us?

Reply:
The legal moves of each piece are formed in this way. For each legal move. There will be a orientation and slot. So basically, it just tells us all the possible ways that a piece can be inserted into the board configuration.

When you do the heuristic.evaluate(s). The piece is not inserted onto the board. So basically, what you are evaluating is the board configuration without the piece on the board.

Instead, we should evaluate the board configuration after the piece is dropped onto the board. So since we know what are all the possible moves of a piece (int [][] legalMoves), we will test each move for the piece and drop the piece (with that orientation in that given slot) onto the board to get the new board configuration with the piece on it, then using the heuristic.evaluate(s) to get the heuristic score.

Thats what I think